pkg/apis/eventing/v1alpha1/cluster_provisioner_types.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"encoding/json"
+	"github.com/knative/pkg/apis"
+	"github.com/knative/pkg/webhook"
+)
+
+// +genclient
+// +genclient:noStatus
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ClusterProvisioner encapsulates a provisioning strategy for the backing
+// resources required to realize a particular resource type.
+type ClusterProvisioner struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the Types provisioned by this Provisioner.
+	Spec ClusterProvisionerSpec `json:"spec"`
+
+	// Status is the current status of the Provisioner.
+	// +optional
+	Status ClusterProvisionerStatus `json:"status,omitempty"`
+}
+
+// Check that ClusterProvisioner can be validated and can be defaulted.
+var _ apis.Validatable = (*ClusterProvisioner)(nil)
+var _ apis.Defaultable = (*ClusterProvisioner)(nil)
+var _ runtime.Object = (*ClusterProvisioner)(nil)
+var _ webhook.GenericCRD = (*ClusterProvisioner)(nil)
+
+// ClusterProvisionerSpec is the spec for a ClusterProvisioner resource.
+type ClusterProvisionerSpec struct {
+	// TODO: Generation does not work correctly with CRD. They are scrubbed
+	// by the APIserver (https://github.com/kubernetes/kubernetes/issues/58778)
+	// So, we add Generation here. Once that gets fixed, remove this and use
+	// ObjectMeta.Generation instead.
+	// +optional
+	Generation int64 `json:"generation,omitempty"`
+
+	// Realizes is the kind of the resource the controller watches to produce
+	// required  backing resources.
+	// +required
+	Realizes metav1.GroupKind `json:"kind"`
```

Comment:
Nice find, this is in fact what I was looking for.

Reply:
I'm assuming that we changed the name so that we didn't have a "Kind.Kind" stutter?

What about `ResourceType` or `Reconciles`?